common/src/KokkosFFT_Helpers.hpp

```diff
@@ -377,167 +130,82 @@ struct Roll<ExecutionSpace, ViewType, 6, iType> {
-/// \tparam ViewType The type of the Kokkos View.
-/// \tparam iType The index type used for the view.
-template <typename ExecutionSpace, typename ViewType, typename iType>
-struct Roll<ExecutionSpace, ViewType, 7, iType> {
-  using ArrayType          = Kokkos::Array<std::size_t, 7>;
-  using LayoutType         = typename ViewType::array_layout;
-  using ManageableViewType = typename manageable_view_type<ViewType>::type;
-
-  using policy_type = Kokkos::MDRangePolicy<
-      ExecutionSpace,
-      Kokkos::Rank<6, Kokkos::Iterate::Default, Kokkos::Iterate::Default>,
-      Kokkos::IndexType<iType>>;
-
-  ViewType m_x;
-  ManageableViewType m_tmp;
-  ArrayType m_shifts;
 
-  /// \brief Constructor for the Roll functor.
+  /// \brief Helper functor to perform the roll operation
   ///
-  /// \param x[in,out] The input/output Kokkos view to be shifted.
-  /// \param shifts[in] The shift amounts for each axis.
-  /// \param exec_space[in] The Kokkos execution space to be used (defaults to
-  /// ExecutionSpace()).
-  Roll(const ViewType& x, const ArrayType& shifts,
-       const ExecutionSpace exec_space = ExecutionSpace())
-      : m_x(x),
-        m_tmp("tmp", create_layout<LayoutType>(extract_extents(x))),
-        m_shifts(shifts) {
-    iType n0 = m_x.extent(0), n1 = m_x.extent(1), n2 = m_x.extent(2),
-          n3 = m_x.extent(3), n4 = m_x.extent(4), n5 = m_x.extent(5);
-    policy_type policy(exec_space, {0, 0, 0, 0, 0, 0}, {n0, n1, n2, n3, n4, n5},
-                       {4, 4, 4, 1, 1, 1});
-    Kokkos::parallel_for("KokkosFFT::roll-7D", policy, *this);
-    Kokkos::deep_copy(exec_space, m_x, m_tmp);
-  }
-
-  KOKKOS_INLINE_FUNCTION
-  void operator()(iType i0, iType i1, iType i2, iType i3, iType i4,
-                  iType i5) const {
-    auto get_dst = [&](iType idx_src, std::size_t axis) {
-      return (idx_src + iType(m_shifts[axis])) % iType(m_x.extent(axis));
-    };
-
-    iType i0_dst = get_dst(i0, 0);
-    iType i1_dst = get_dst(i1, 1);
-    iType i2_dst = get_dst(i2, 2);
-    iType i3_dst = get_dst(i3, 3);
-    iType i4_dst = get_dst(i4, 4);
-    iType i5_dst = get_dst(i5, 5);
-    for (iType i6 = 0; i6 < iType(m_x.extent(6)); i6++) {
-      iType i6_dst = get_dst(i6, 6);
-      m_tmp(i0_dst, i1_dst, i2_dst, i3_dst, i4_dst, i5_dst, i6_dst) =
-          m_x(i0, i1, i2, i3, i4, i5, i6);
+  /// \tparam IndexSequence The index sequence used for the parallel execution.
+  template <typename IndexSequence>
+  struct RollInternal;
+
+  template <std::size_t... Idx>
+  struct RollInternal<std::index_sequence<Idx...>> {
+    template <std::size_t I>
+    using IndicesType = iType;
+
+    ViewType m_x;
+    ManageableViewType m_buffer;
+    ArrayType m_shifts;
+
+    RollInternal(const ViewType& x, const ManageableViewType& buffer,
+                 const ArrayType& shifts)
+        : m_x(x), m_buffer(buffer), m_shifts(shifts) {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const IndicesType<Idx>... indices) const {
```

Comment:
I think you can just template the `operator()` instead of templating the struct, since you already cast to `iType` it's ok to get rid of `IndicesType` here
```suggestion
  struct RollInternal {
    ViewType m_x;
    ManageableViewType m_buffer;
    ArrayType m_shifts;

    RollInternal(const ViewType& x, const ManageableViewType& buffer,
                 const ArrayType& shifts)
        : m_x(x), m_buffer(buffer), m_shifts(shifts) {}

    template<typename... IndicesType>
    KOKKOS_INLINE_FUNCTION
    void operator()(const IndicesType... indices) const {
```

Reply:
That is certainly nicer. I have updated accordingly.